File: frontend/src/constants/fieldLimits.ts
export const FIELD_LIMITS = {
  personName: 100,
  position: 100,
  adminUsername: 100,
  password: 100,
  room: 100,
  schedule: 100,
  structureName: 150,
  email: 150,
  title: 200,
  address: 200,
  url: 255,
} as const;

/** Строга валідація email: вимагає коректний домен (напр. @gmail.com, @lnu.edu.ua) */
export const EMAIL_PATTERN = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
export const EMAIL_INPUT_PATTERN = '[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}';
export const EMAIL_HINT = 'Введіть коректний email у форматі name@domain.com';

export const isValidEmail = (value: string) => EMAIL_PATTERN.test(value.trim());

export const DUPLICATE_NAME_MESSAGE = "Ця назва вже використовується. Оберіть іншу.";

export function isDuplicateStructureName(
  name: string | undefined,
  items: { id: number; name: string }[],
  editingId?: number | null,
): boolean {
  const normalized = name?.trim().toLowerCase();
  if (!normalized) return false;

  return items.some((item) => {
    if ((item.name || "").trim().toLowerCase() !== normalized) return false;
    if (editingId == null) return true;
    return Number(item.id) !== Number(editingId);
  });
}

File: backend/Backend/Utils/FieldLimits.cs
using System.Text.RegularExpressions;

namespace ProfkomBackend.Utils
{
    public static class FieldLimits
    {
        public const int PersonName = 100;
        public const int Position = 100;
        public const int AdminUsername = 100;
        public const int Password = 100;
        public const int Room = 100;
        public const int Schedule = 100;
        public const int StructureName = 150;
        public const int Email = 150;
        public const int Title = 200;
        public const int Address = 200;
        public const int Url = 255;

        /// <summary>Строга валідація email: вимагає коректний домен (напр. @gmail.com, @lnu.edu.ua)</summary>
        public const string EmailPattern = @"^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$";
        public const string EmailFormatMessage = "Введіть коректний email у форматі name@domain.com";

        public static bool IsValidEmail(string? value)
        {
            if (string.IsNullOrWhiteSpace(value)) return false;
            return Regex.IsMatch(value.Trim(), EmailPattern);
        }
    }
}
